File: AlgorithmStudy/src/y2022/jul/Solution0728.java
package y2022.jul;

// https://leetcode.com/problems/valid-anagram/
public class Solution0728 {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;

        int[] alphabet = new int[26];

        for (int i = 0; i < s.length(); i++) {
            alphabet[s.charAt(i) - 'a']++;
            alphabet[t.charAt(i) - 'a']--;
        }

        for (int i = 0; i < alphabet.length; i++) {
            if (alphabet[i] != 0) {
                return false;
            }
        }

        return true;
    }
}

File: AlgorithmStudy/src/y2023/jul/Solution0728.kt
package y2023.jul

// https://leetcode.com/problems/predict-the-winner/
class Solution0728 {
    fun PredictTheWinner(nums: IntArray): Boolean {
        val n = nums.size
        val dp = Array(n) { IntArray(n) }
        for (i in 0 until n) {
            dp[i][i] = nums[i]
        }

        for (diff in 1 until n) {
            for (left in 0 until n - diff) {
                val right = left + diff
                dp[left][right] = Math.max(nums[left] - dp[left + 1][right],
                        nums[right] - dp[left][right - 1])
            }
        }

        return dp[0][n - 1] >= 0
    }
}